### ISSUE TITLE
How is y_pred conneted to output of generator implicitly in keras?

### ISSUE THREAD

**Issue by Rafeal23** [breizhn/DTLN#13]
Hi, thanks for sharing the code!

I am new to Keras. I am try to train your model but confused about how y_pred is generated.

The input of model is yield by the create_generator function

            yield in_dat.astype('float32'), tar_dat.astype('float32')

then tf_data_set is created from the generator

     self.tf_data_set = tf.data.Dataset.from_generator(
                    self.create_generator,
                    (tf.float32, tf.float32),
                    output_shapes=(tf.TensorShape([self.len_of_samples]), \
                                   tf.TensorShape([self.len_of_samples])),
                    args=None
                    )

the data batches are generated by .batch op

    generator_input = audio_generator(path_to_train_mix, 
                                      path_to_train_speech, 
                                      len_in_samples, 
                                      self.fs, train_flag=True)
    dataset = generator_input.tf_data_set
    dataset = dataset.batch(self.batchsize, drop_remainder=True).repeat()
    # calculate number of training steps in one epoch
    steps_train = generator_input.total_samples//self.batchsize
    # create data generator for validation data
    generator_val = audio_generator(path_to_val_mix,
                                    path_to_val_speech, 
                                    len_in_samples, self.fs)
    dataset_val = generator_val.tf_data_set
    dataset_val = dataset_val.batch(self.batchsize, drop_remainder=True).repeat()

then modeo.fir is called to fit the model

    self.model.fit(
        x=dataset, 
        batch_size=None,
        steps_per_epoch=steps_train, 
        epochs=self.max_epochs,
        verbose=1,
        validation_data=dataset_val,
        validation_steps=steps_val, 
        callbacks=[checkpointer, reduce_lr, csv_logger, early_stopping],
        max_queue_size=50,
        workers=4,
        use_multiprocessing=True)

In terms of calculation of the loss

        loss = tf.squeeze(self.cost_function(y_pred,y_true))
        # calculate mean over batches
        loss = tf.reduce_mean(loss)

y_pred,y_true is passed to the cost_function.

For some other codes that I've read, usually y_pred and y_true is clearly calculated, such as:

y_pred = Model(x_train_data) # output of model
y_true = x_label

I know that here, y_pred is the output of the model when in_dat is inputted to the model, and y_true is tar_dat. However, in yout model, I cannot find such calculations.

To summarize, my questions are:

1) How is y_pred conneted to output of generator (in_data)?
2) If I want to modify ouput of the generator and the loss function(e.g. original noise data and filenames), what should I do to differentiate which data are input to the model(such as in_dat) and which are not (such as tar_dat)?

**Comment by breizhn** [comment]
Hi Eric,

> * How is y_pred conneted to output of generator (in_data)?

The `model.fit` fuction is passing the output of the generator to the model and generates `y_pred`. After that y_pred is passed to the cost function, which already compiled into the model with `model.compile`. The people calling the model for predictions probably implemented a custom training loop. For that see the Tensorflow [documentation](https://www.tensorflow.org/tutorials/customization/custom_training). In conclusion, if you don't wont to apply fancy cost functions or not standard gradient clipping, `model.fit` is the easiest way to train a model.

> * If I want to modify ouput of the generator and the loss function(e.g. original noise data and filenames), what should I do to differentiate which data are input to the model(such as in_dat) and which are not (such as tar_dat)?

For the loss function you can use the current loss function as template. The cost function always gets `y_pred,y_true` as input. Try to play a bit around with my generator. The Dataset is wrapper for the yield function. And for any additional information, the tensorflow documentation is very helpful for this kind of stuff.

I hope, this answers your questions a bit.

**Comment by Rafeal23** [comment]
Hi Nils,
Thank you for your reply. I understand much better now!
